pepsico/app_calc.py

```diff
@@ -0,0 +1,86 @@
+import xarray as xr
+
+
+#This is what we should need for the app
+#For the map:
+#-- read_data for histo and 1 model
+#-- seasonal_data for specific years range
+#-- seasanal_data.mean(dim="T") - seasonal_histo.mean(dim="T")
+#-- unit_conversion
+#-- then maybe some other things to beautify map tbd
+#
+#For the ts:
+#-- read_data for all histo and scenarios
+#-- select X and Y
+#-- seasonal_data
+#-- append or figure out what the graph will need as input to plot histo
+#followed by different scenarios
+#-- unit_conversion
+#-- then maybe some other things to beautify map tbd
+
+
+def read_data(scenario, model, variable):
+
+    return xr.open_zarr(
+        f'/Data/data24/ISIMIP3b/InputData/climate/atmosphere/bias-adjusted/global'
+        f'/monthly/{scenario}/{model}/zarr/{variable}'
+    )[variable]
+
+
+def seasonal_data(monthly_data, start_month, end_month, start_year=None, end_year=None):
+
+    #NDF and DJF are considered part of the year of the 1st month
+    if ((end_year != None) and (start_month > end_month)):
+        end_year = str(int(end_year) + 1)
+    #Reduce data size
+    monthly_data = monthly_data.sel(T=slice(start_year, end_year))
+    #Find edges of seasons
+    start_edges = monthly_data["T"].where(
+        lambda x: x["T"].dt.month == start_month, drop=True,
+    )
+    end_edges = monthly_data["T"].where(
+        lambda x: x["T"].dt.month == end_month, drop=True,
+    )
+    #Select data and edges to avoid partial seasons at the edges of the edges
+    monthly_data = monthly_data.sel(T=slice(start_edges[0], end_edges[-1]))
+    start_edges = start_edges.sel(T=slice(start_edges[0], end_edges[-1]))
+    end_edges = (end_edges
+        .sel(T=slice(start_edges[0], end_edges[-1]))
+        .assign_coords(T=start_edges["T"])
```

Comment:
We need 2 items to describe the season. Can be center and width, or start and end. Center seems more appropriate if you plot a time series for instance that the point is located at the center rather than an edge; or if thinking of operations like derivative/integral, then the centers become edges and vice-versa. The advantage also is that width will be a constant (as opposed to a vector) for regular cases, and can be equal to the data step. Center and width is also how we naturally interpret gridded data for lon and lat, where the width is inferred implictely by the step (or resolution). So center and width seems to be a more general case... And with DataArrays accepting several coordinates, width as a vector could be dealt with (which is not the case in Ingrid).

But I am biased by 15 years thinking that way...

Reply:
Yes, this is the argument I remember you making in the past, which is why it struck me as strange that you're setting T to the start of the season here. Or have I misinterpreted something?

Reply:
I set it as we did for the already written daily-input case... I am not sure why that choice was made then. The concept of width would need more work with Xarray though... It doesn't really have one; and it doesn't really have that much of a concept of a month, other than by labelling I guess, since timedelta doesn't know what a month is either (I guess because that's when partition of time stops to be regular...)

Reply:
Right, xarray doesn't have any built-in support for cell widths, but we can create datasets with cell width metadata, and we can write library functions that use that metadata.